### PR TITLE
Refine imperative VarBase

### PR DESCRIPTION
**Pull request by velconia**

### CODE REVIEW

**Comment by luotao1** [comment]
Could you fix the warning in 
```
[04:17:17][Step 1/1] /paddle/paddle/fluid/imperative/layer.cc: In function 'void paddle::imperative::AddTo(paddle::framework::Variable*, paddle::framework::Variable*)':
[04:17:17][Step 1/1] /paddle/paddle/fluid/imperative/layer.cc:45:24: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[04:17:17][Step 1/1]    for (size_t i = 0; i < src_tensor->numel(); ++i) {
[04:17:17][Step 1/1]          
```
http://ci.paddlepaddle.org/viewLog.html?buildId=45847&tab=buildLog&buildTypeId=Paddle_PrCi&logTab=tree&filter=all&_focus=1125#_state=93

**Comment by velconia** [comment]
OK, I will fix the warning in this PR

File: paddle/fluid/imperative/layer.h [hunk 1]
```diff
@@ -115,12 +123,16 @@ class VarBase {
   int pre_op_out_idx_;
 
   framework::VarDesc* var_desc_;
-  framework::Variable* var_;
-  framework::Variable* grads_;
+
+  std::shared_ptr<framework::Variable> var_;
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.h]: delete in destructor and don't use share_ptr now?

File: paddle/fluid/imperative/layer.h [hunk 2]
```diff
@@ -115,12 +123,16 @@ class VarBase {
...
+  std::shared_ptr<VarBase> grads_;
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.h]: same here.

File: paddle/fluid/imperative/tracer.cc [hunk 3]
```diff
@@ -15,5 +15,146 @@
 #include "paddle/fluid/imperative/tracer.h"
 
 namespace paddle {
-namespace imperative {}  // namespace imperative
+namespace imperative {
+
+void CreateGradOp(const framework::OpDesc& op_desc,
```
Comment (panyx0718) [inline paddle/fluid/imperative/tracer.cc:20]: any thing changed here? Or just moved to here?

Comment (velconia) [reply paddle/fluid/imperative/tracer.cc:20]: Just moved to here.

File: paddle/fluid/imperative/layer.h (hunk 2 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.h]: Done

File: paddle/fluid/imperative/layer.h (hunk 1 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.h]: Done

**Review by panyx0718 — DISMISSED** [review]

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**